README.adoc
= `test161`: A Testing Tool for OS/161

`test161` is a command line tool for automated testing of
http://os161.eecs.harvard.edu[OS/161] instructional operating system kernels
that run inside the `sys161` (System/161)
https://en.wikipedia.org/wiki/R3000[MIPS R3000] simulator. You are probably
not interested in `test161` unless you are a student taking or an instructor
teaching a course that uses OS/161.

== Installing Go

Many Linux distributions package fairly out-of-date versions of Go. Instead,
we encourage you to install the https://github.com/moovweb/gvm[Go Version Manager (GVM)]:

[source,bash]
----
sudo apt-get install -y curl bison # Install requirement
bash < <(curl -s -S -L https://raw.githubusercontent.com/moovweb/gvm/master/binscripts/gvm-installer)
source $HOME/.gvm/scripts/gvm
----

At this point you are ready to start using GVM. We are currently building and
testing `test161` with Go version `go1.5.3`. However, because the Go compiler
is now written in Go, installing versions of Go past 1.5 require install Go
version 1.4 first.

[source,bash]
----
gvm install go1.4
gvm use 1.4
gvm install go1.5.3
gvm use 1.5.3 --default
----

=== `GOPATH`

Note that `gvm` will set your `GOPATH` and `PATH` variables properly to allow
you to run Go binaries that you install. However, if you are interested in
writing Go code you should set a more accessible `GOPATH` as described as https://golang.org/doc/code.html#GOPATH[described
here.]

== Installing `test161`

Once you have Go installed, upgrading or installing `test161` is simple:

[source,bash]
----
go get -u github.com/ops-class/test161/test161
----

== Configuration

In order to run `test161`, it needs to know about which <<tests>>, 
<<targets>>, and <<commands>> it can run, and what to expect from  each command.
Additionally, `test161` needs to know where to find your root directory for
testing and your repository for submission. This information is specified in your
`.test161.conf` file, which can be located in your `$HOME` directory, or whichever
directory you plan to run `test161` in. If you run `test161 run` without a config
file set up, an example will be displayed on screen.  You can redirect this output
to `.test161.conf` and edit it accordingly.

[source,bash]
----
test161 run > ~/.test161.conf
vim ~/.test161.conf
----

== Usage

=== Running Tests

To run a tests, group of tests, or target, use the `test161 run <names>` sub-command.
Here, names can be a single target, a list of tests, or a list of <<tags>>.footnote:[In the case
that tag and target names conflict, specify `-tag` if you mean tag.] For test files,
<names> is a list of http://www.linuxjournal.com/content/globstar-new-bash-globbing-option [globstar]
style file names.  The following are all valid commands:

[source,bash]
----
test161 run sync/*.t         # Run all tests in the tests/sync
test161 run **/l*.t          # Run all tests in all sub-directories beginning with 'l'
test161 run syncprobs/sp*.t  # Run the syncprobs
test161 run sync/lt1.t       # Run lock test 1
test161 run locks            # Run all lock tests (tests tagged with 'locks')
test161 run asst1            # Run the asst1 target
----

==== Command Line Flags
There are several command line flags that can be specified to customize how `test161` runs tests.

* `-dry-run` (`-r`): Show the tests that would be run, but don't run them.
* `-sequential` (`-s`): By default the output of all tests are interleaved, which can be hard to debug.
Specify this option to run tests one at a time.
* `-no-dependencies` (`-n`): Run the given tests without also running their dependencies. In test files, you can specify a
test's dependencies, and the test will be skipped if the dependency fails. The default behevior is to run all dependencies
along with each test specified on the command line. This option does not apply for targets.
* `-verbose` (`-v`): There are 3 levels of output: `loud` (default), `quiet` (no test output), and `whisper` (only final summary, no
per test status).

== Requirements

* `sys161` and `disk161` in the path.

== Default Settings

....
conf:
  cpus: 8
  ram: 1M
  disk1:
    sectors: min(2 * ram, 8000) # 8000 is a minimum due to a current sys161 bug
    rpm: 7200
    nodoom: true
  disk2: # disabled by default, but uses these defaults if configured
    sectors: min(2 * ram, 8000)
    rpm: 7200
    nodoom: false
  random: seed=random # random number generated at configuration time
stat:
  resolution: 0.01
  window: 1
monitor:
  enabled: "true
  intervals: 10
  kernel:
    enabled: false
    min: 0.001
    max: 1.0
  user:
    enabled: false
    min: 0.0001
    max: 1.0
  timeouts:
    prompt: 300
    progress: 10
....

== Commands, Tests, and Targets

=== [[commands]]Commands

The basic unit in `test161` is a command. such as `lt1` for running Lock Test 1,
or `sp1` to run the whalemating test.  Information about what to
expect when running these commands, as well as what input/ouput they expect
 is specified in the `commands/` directory in your test161 root directory.
All .tc (test command) file in this directory will be loaded and commands must
only be specified once.

=== [[tests]]Tests

Test files (`*.t`) are located in the `tests/` directory in your test161 root
directory. This directory can contain subdirectories to help organize tests.
Each test consists of one or more commands, and each test can have its own
`sys161` configuration.  Tests are run in their own sandboxed envrionment, 
but commands within the test are executed within the same `sys161` session.
Some tests will consist of multiple commands, and such tests are designed to
stress test your system.

=== [[targets]]Targets

Target files (`*.tt`) are located in the `targets/` directory in your test161 root
directory. Targets specify which tests are run for each assignment, and
how the scoring is distributed. When you `test161 submit` your assignments, you will
specify which target to submit to.

== Features

=== Testfile Syntactic Sugar

A line starting with `$` will be run in the shell and start the shell as
needed. Lines not starting with `$` are run from the kernel prompt and get
there if necessary by exiting the shell. `sys161` shuts down cleanly without
requiring the test manually exit the shell and kernel, as needed.

So this test:
....
$ /bin/true
....

Expands to:
....
s
/bin/true
exit
q
....

*Note that commands run in the shell _must_ be prefixed with `$`.* Otherwise
`test161` will consider them a kernel command and exit the shell before
running them. For example:

This test is probably not what you want:
....
s
/bin/true
....

Because it will expand to:
....
s
exit
/bin/true # not a kernel command
....

But this is so much simpler, right?
....
$ /bin/true
....

=== [[tags]]Test Tags
Optionally, tests can have one or more tags. `test161` can be invoked to run these tests
as a group with `test161 run <tag>`.

=== Progress Tracking Using `stat161` Output

`test161` uses the collected `stat161` output produced by the running kernel to
detect deadlocks, livelocks, and other forms of stalls. We do this using
several different strategies:

. *Progress and prompt timeouts.* Testfiles can configure both progress
(`monitorconf.timeouts.progress`) and prompt (`monitorconf.timeouts.prompt`)
timeouts. The former is used to kill the test if no output has appeared, while
the latter is passed to `expect` and used to kill the test of the prompt is
delayed. Ideally OS/161 tests should produce some output while they run to
help keep the progress timeout from firing, but the other progress tracking
strategies described below should also help.
. *User and kernel maximum and minimum cycles.* `test161` maintains a buffer
of statistics over a configurable number of `stat161` intervals. Limits on the
minimum and maximum number of kernel and user cycles (expressed as fractions)
over this buffer can help detect deadlocks (minimum) and livelocks (maximum).
User limits are only applied when running in userspace.
.  Note that `test161`
also checks to ensure that there are no user cycles generated when we are
running in kernel mode, which could be caused by a hung progress.


=== Running multiple tests and dependencies

=== Correctness vs. Grading

=== Security

=== Multiple output strategies

`test161` supports different output strategies through its PersistenceManager
interface. Each TestEnvironment as a PersistenceManager which receives
callbacks when events happen, like when scores changes, status change, or when
output lines are added. This allows multiple implementations to handle output
as they wish. The test161 client utility implements the interface through
its ConsolePersistence type, which writes all input to stdout. The server uses
a MongoPersistence type which outputs JSON data to our mongo backend server.


== TODOs

=== Nits

* Replace "0 bytes" with "None!" in the memory leak output for extra
motivation :-).
* Colored test output on terminals that support it? (Particularly for correct/incorrect.)
* Either don't require a `sys161.conf` file or print a better error message
when it is missing.
* Support for environments that don't have `.root` or `.src`. Probably just
don't require these, but this will mean that you have to run `test161 run ...`
from your root directory and `test161 submit` from your source directory.
* Handle missing newline correctly. Test with shll for lossy shell support.
* sys161 version checks
* Order the test output in some meaningful way, probably by depth in the
dependency graph. (That way all skipped tests should be shown last.)
** Not necessarily true. You could have a long, unrelated branch that
succeeds (even leaf nodes), but some unrelated depencency fails early. I made the
default print order topological sort, but it's still confusing. I added the
reason a test is skipped, which helps. Maybe a nice ASCII art tree would work here...
* Better error messages and hints about repository problems:
** Check and fail if it has inappropriate files (`.o`), or is too large. 
(Prevent backend storage DOS attacks.)
* Add support for targets and tests to the bash completion script

=== More Syntactic Sugar

Nice to add "|" as a prefix to wrap a command with memory leak detection. Need
to also process other prefixes in this case so that:

....
| cvt1
....

expands to:
....
khu
cvt1
khu
....

and:
....
|$ /bin/true
....

expands to:
....
khu
s
/bin/true
exit
khu
....

Also it would be useful to add a multiplier syntax so that:
....
4x cvt1
....

expands to:
....
cvt1
cvt1
cvt1
cvt1
....

=== Parallel Testing Output

It would be cool to be able to print serial output from one test while queuing
the output from other tests. Maybe using curses to maintain a line at the end
of the screen showing the tests that are being run.

=== Tag Descriptions and Querying

It would be nice to be able to add descriptions to tags and have `test161`
print all tests that fall into a tag (or target) along with the description.

=== Key saving

Now that we are having students save a key through the web interface we need
to make sure that these keys get saved, associated with each successful
submission, and not destroyed even if they are changed later.

=== Configuration override

It would be great if `test161 run boot.t --sys161-cpus=1` worked properly. I
think that there is a library for this.

=== Support for GDB backtraces on error

It should be possible to automate the process of hooking up a debugger and
running BT on panics.

=== Server Binary
* Moving window for stats API


debian/changelog
test161 (1.2.0) trusty; urgency=medium

  * Initial release.

 -- Geoffrey Challen <geoffrey.challen@gmail.com>  Thu, 03 Mar 2016 08:22:06 -0500


conf.go
package test161

import (
	"bytes"
	"errors"
	"fmt"
	"github.com/ericaro/frontmatter"
	"github.com/imdario/mergo"
	"github.com/satori/go.uuid"
	"io/ioutil"
	"math/rand"
	"reflect"
	"regexp"
	"strings"
	"sync"
	"text/template"
)

// Many of the values below that come in from YAML are string types. This
// allows us to work around Go not having a nil type for things like ints, and
// also allows us to accept values like "10M" as numbers. Non-string types are
// only used when the default doesn't make sense. Given that the ultimate
// destination of these values is strings (sys161.conf configuration or JSON
// output) it doesn't matter.

type Sys161Conf struct {
	Path   string   `yaml:"path" json:"path"`
	CPUs   uint     `yaml:"cpus" json:"cpus"`
	RAM    string   `yaml:"ram" json:"ram"`
	Disk1  DiskConf `yaml:"disk1" json:"disk1"`
	Disk2  DiskConf `yaml:"disk2" json:"disk2"`
	Random uint32   `yaml:"-" json:"randomseed" bson:"randomseed"`
}

type DiskConf struct {
	Enabled string `yaml:"enabled" json:"enabled"`
	RPM     uint   `yaml:"rpm" json:"rpm"`
	Bytes   string `yaml:"bytes" json:"bytes"`
	NoDoom  string `yaml:"nodoom" json:"nodoom"`
}

type StatConf struct {
	Resolution float32 `yaml:"resolution" json:"resolution"`
	Window     uint    `yaml:"window" json:"window"`
}

type MonitorConf struct {
	Enabled         string  `yaml:"enabled" json:"enabled"`
	Window          uint    `yaml:"window" json:"window"`
	Kernel          Limits  `yaml:"kernel" json:"kernel"`
	User            Limits  `yaml:"user" json:"user"`
	ProgressTimeout float32 `yaml:"progresstimeout" json:"progresstimeout"`
}

type Limits struct {
	EnableMin string  `yaml:"enablemin" json:"enablemin"`
	Min       float64 `yaml:"min" json:"min"`
	Max       float64 `yaml:"max" json:"max"`
}

type MiscConf struct {
	CommandRetries   uint    `yaml:"commandretries" json:"commandretries"`
	PromptTimeout    float32 `yaml:"prompttimeout" json:"prompttimeout"`
	CharacterTimeout uint    `yaml:"charactertimeout" json:"charactertimeout"`
	TempDir          string  `yaml:"tempdir" json:"-" bson:"-"`
	RetryCharacters  string  `yaml:"retrycharacters" json:"retrycharacters"`
	KillOnExit       string  `yaml:"killonexit" json:"killonexit"`
}

type CommandConf struct {
	Prefix string `yaml:"prefix" json:"prefix"`
	Prompt string `yaml:"prompt" json:"prompt"`
	Start  string `yaml:"start" json:"start"`
	End    string `yaml:"end" json:"end"`
}

var CONF_DEFAULTS = Test{
	Sys161: Sys161Conf{
		Path: "sys161",
		CPUs: 8,
		RAM:  "1M",
		Disk1: DiskConf{
			Enabled: "true",
			Bytes:   "4M",
			RPM:     7200,
			NoDoom:  "true",
		},
		Disk2: DiskConf{
			Enabled: "false",
			Bytes:   "2M",
			RPM:     7200,
			NoDoom:  "false",
		},
	},
	Stat: StatConf{
		Resolution: 0.01,
		Window:     1,
	},
	Monitor: MonitorConf{
		Enabled: "true",
		Window:  400,
		Kernel: Limits{
			EnableMin: "false",
			Min:       0.001,
			Max:       1.0,
		},
		User: Limits{
			EnableMin: "false",
			Min:       0.0001,
			Max:       1.0,
		},
		ProgressTimeout: 10.0,
	},
	Misc: MiscConf{
		CommandRetries:   5,
		PromptTimeout:    300.0,
		CharacterTimeout: 1000,
		RetryCharacters:  "true",
		KillOnExit:       "false",
	},
}

func confFromString(data string) (*Test, error) {
	t := new(Test)

	err := frontmatter.Unmarshal([]byte(data), t)
	if err != nil {
		return nil, err
	}
	t.Sys161.Random = rand.Uint32() >> 16

	// TODO: Error checking here

	return t, nil
}

// TestFromFile parses the test file and sets configuration defaults.
func TestFromFile(filename string) (*Test, error) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, fmt.Errorf("Error reading file %v: %v", filename, err)
	}
	test, err := TestFromString(string(data))
	if err != nil {
		err = fmt.Errorf("Error loading test file %v: %v", filename, err)
	}
	return test, err
}

// TestFromFile parses the test string and sets configuration defaults.
func TestFromString(data string) (*Test, error) {

	t, err := confFromString(data)
	if err != nil {
		return nil, err
	}

	// Check for empty commands and expand syntatic sugar before getting
	// started. Doing this first makes the main loop and retry logic simpler.

	t.ID = uuid.NewV4().String()
	err = t.initCommands()
	if err != nil {
		return nil, err
	}

	t.Result = TEST_RESULT_NONE

	return t, nil
}

func (t *Test) MergeConf(defaults Test) error {
	return mergo.Map(t, defaults)
}

const SYS161_TEMPLATE = `0 serial
1	emufs
{{if eq .Disk1.Enabled "true"}}
2	disk rpm={{.Disk1.RPM}} file=LHD0.img {{if eq .Disk1.NoDoom "true"}}nodoom{{end}} # bytes={{.Disk1.Bytes }}
{{end}}
{{if eq .Disk2.Enabled "true"}}
3	disk rpm={{.Disk2.RPM}} file=LHD1.img {{if eq .Disk2.NoDoom "true"}}nodoom{{end}} # bytes={{.Disk2.Bytes }}
{{end}}
28	random seed={{.Random}}
29	timer
30	trace
31	mainboard ramsize={{.RAM}} cpus={{.CPUs}}`

var confTemplate *template.Template
var confOnce sync.Once

// PrintConf formats the test configuration for use by sys161 via the sys161.conf file.
func (t *Test) PrintConf() (string, error) {

	confOnce.Do(func() { confTemplate, _ = template.New("conf").Parse(SYS161_TEMPLATE) })

	buffer := new(bytes.Buffer)

	// Can't fail, even if the test isn't properly initialized, because
	// t.Sys161 matches the template
	confTemplate.Execute(buffer, t.Sys161)

	var confString string
	for _, line := range strings.Split(strings.TrimSpace(buffer.String()), "\n") {
		if strings.TrimSpace(line) != "" {
			confString += line + "\n"
		}
	}
	return confString, nil
}

func (t *Test) confEqual(t2 *Test) bool {
	return t.Sys161 == t2.Sys161 &&
		t.Stat == t2.Stat &&
		t.Monitor == t2.Monitor &&
		t.Misc == t2.Misc &&
		reflect.DeepEqual(t.CommandConf, t2.CommandConf)
}

var prefixRegexp *regexp.Regexp
var prefixOnce sync.Once

func splitPrefix(commandLine string) (string, string) {
	prefixOnce.Do(func() { prefixRegexp = regexp.MustCompile(`^([!@#$%^&*]) `) })
	matches := prefixRegexp.FindStringSubmatch(strings.TrimSpace(commandLine))
	if len(matches) == 0 {
		return "", strings.TrimSpace(commandLine)
	} else {
		return matches[1], strings.TrimSpace(commandLine[1:])
	}
}

func (t *Test) checkCommandConf() error {
	prefixes := "$"
	errorMsg := ""
	var commandConf CommandConf
	for _, commandConf = range t.CommandConf {
		if commandConf.Prefix == "" || commandConf.Prompt == "" ||
			commandConf.Start == "" || commandConf.End == "" {
			errorMsg = "need to specific command prefix, prompt, start, and end"
			break
		}
		if len(commandConf.Prefix) > 1 {
			errorMsg = "illegal multicharacter prefix"
			break
		}
		prefix, _ := splitPrefix(commandConf.Prefix + " t")
		if prefix == "" {
			errorMsg = "invalid prefix"
			break
		} else if prefix == "$" {
			errorMsg = "the $ prefix is reserved for the shell"
			break
		} else if strings.ContainsAny(prefixes, prefix) {
			errorMsg = "duplicate prefix"
			break
		}
		prefixes += prefix
		startPrefix, _ := splitPrefix(commandConf.Start)
		if startPrefix != "" && startPrefix == prefix {
			errorMsg = "command start cannot start with own prefix"
			break
		}
		endPrefix, _ := splitPrefix(commandConf.End)
		if endPrefix != "" {
			errorMsg = "command exits should not contain a prefix"
			break
		}
	}
	if errorMsg == "" {
		for _, commandConf = range t.CommandConf {
			startPrefix, _ := splitPrefix(commandConf.Start)
			if startPrefix != "" && !strings.ContainsAny(prefixes, startPrefix) {
				errorMsg = "command start contains an unknown prefix"
				break
			}
		}
	}
	if errorMsg != "" {
		return errors.New(fmt.Sprintf("test161: %s (%v)", errorMsg, commandConf))
	} else {
		return nil
	}
}

var KERNEL_COMMAND_CONF = &CommandConf{
	Prompt: `OS/161 kernel [? for menu]: `,
	End:    "q",
}
var SHELL_COMMAND_CONF = &CommandConf{
	Prefix: "$",
	Prompt: `OS/161$ `,
	Start:  "s",
	End:    "exit",
}

func (t *Test) commandConfFromLine(commandLine string) (string, *CommandConf) {
	prefix, commandLine := splitPrefix(commandLine)
	if prefix == "" {
		return commandLine, KERNEL_COMMAND_CONF
	} else if prefix == "$" {
		return commandLine, SHELL_COMMAND_CONF
	} else {
		for _, commandConf := range t.CommandConf {
			if commandConf.Prefix == prefix {
				return commandLine, &commandConf
			}
		}
	}
	return commandLine, nil
}

// Don't get stuck in an infinite loop

const MAX_EXPANSION_LOOPS = 1024

func (t *Test) initCommands() (err error) {

	// Check defined command prefixes
	err = t.checkCommandConf()
	if err != nil {
		return err
	}

	// Set up the command configuration stack
	var commandConfStack []*CommandConf
	commandConfStack = append(commandConfStack, KERNEL_COMMAND_CONF)

	// Set the boot command
	t.Commands = append(t.Commands, &Command{
		Type:          "kernel",
		ID:            uuid.NewV4().String(),
		Test:          t,
		PromptPattern: regexp.MustCompile(regexp.QuoteMeta(KERNEL_COMMAND_CONF.Prompt)),
		Input: InputLine{
			Line: "boot",
		},
		Status:   COMMAND_STATUS_NONE,
		Panic:    CMD_OPT_NO,
		TimesOut: CMD_OPT_NO,
		Timeout:  0.0,
	})

	// Set all confs including kernel and shell
	allConfs := append(t.CommandConf, *SHELL_COMMAND_CONF, *KERNEL_COMMAND_CONF)

	commandLines := strings.Split(strings.TrimSpace(t.Content), "\n")
	for i := 0; i <= MAX_EXPANSION_LOOPS; i++ {
		if i == MAX_EXPANSION_LOOPS {
			return errors.New("test161: infinite loop expanding command list")
		}
		if len(commandConfStack) == 0 {
			if len(commandLines) != 0 {
				return errors.New("test161: premature exit in command list")
			}
			// This is normal exit
			break
		}
		currentConf := commandConfStack[0]

		// May need to clean up previous configurations
		if len(commandLines) == 0 {
			commandLines = []string{strings.TrimSpace(currentConf.Prefix + " " + currentConf.End)}
			continue
		}

		// Peek at the first command
		commandLine := strings.TrimSpace(commandLines[0])
		if commandLine == "" {
			return errors.New("test161: found empty command")
		}

		commandLine, commandConf := t.commandConfFromLine(commandLine)
		if commandConf == nil {
			return errors.New(fmt.Sprintf("test161: command with invalid prefix %v", commandLine))
		}

		if !reflect.DeepEqual(currentConf, commandConf) {
			found := false
			for _, search := range commandConfStack {
				if commandConf == search {
					found = true
					break
				}
			}
			if found {
				commandLines = append([]string{strings.TrimSpace(currentConf.Prefix + " " + currentConf.End)}, commandLines...)
			} else {
				commandLines = append([]string{commandConf.Start}, commandLines...)
			}
			continue
		} else {
			typeConf := currentConf
			nextConf := currentConf
			if commandLine == currentConf.End {
				// The command exits the current configuration
				commandConfStack = commandConfStack[1:]
				if len(commandConfStack) > 0 {
					nextConf = commandConfStack[0]
				} else {
					nextConf = nil
				}
			} else {
				for _, search := range allConfs {
					if search.Start == strings.TrimSpace(currentConf.Prefix+" "+commandLine) {
						// The command starts a new configuration
						commandConfStack = append([]*CommandConf{&search}, commandConfStack...)
						typeConf = commandConfStack[0]
						nextConf = commandConfStack[0]
						break
					}
				}
			}
			var commandType string
			if typeConf.Prefix != "" || strings.HasPrefix(commandLine, "p ") {
				commandType = "user"
			} else {
				commandType = "kernel"
			}
			var promptPattern *regexp.Regexp
			if nextConf != nil {
				promptPattern = regexp.MustCompile(regexp.QuoteMeta(nextConf.Prompt))
			}
			t.Commands = append(t.Commands, &Command{
				Type:          commandType,
				ID:            uuid.NewV4().String(),
				Test:          t,
				PromptPattern: promptPattern,
				Input: InputLine{
					Line: commandLine,
				},
				Status:   COMMAND_STATUS_NONE,
				Panic:    CMD_OPT_NO,
				TimesOut: CMD_OPT_NO,
				Timeout:  0.0,
			})
			commandLines = commandLines[1:]
		}
	}

	return nil
}
